File: src/Runtime/ikChainInfo.ts
import { Quaternion, Vector3 } from "@babylonjs/core/Maths/math.vector";

/**
 * Ik chain information for solving IK
 */
export class IkChainInfo {
    /**
     * Local rotation of the bone for solving IK
     */
    public readonly localRotation: Quaternion;

    /**
     * Local position of the bone for solving IK
     */
    public readonly localPosition: Vector3;

    /**
     * The rotation offset value to be moved by the IK solver
     */
    public readonly ikRotation: Quaternion;

    /**
     * Creates IK chain information
     */
    public constructor() {
        this.localRotation = Quaternion.Identity();
        this.localPosition = Vector3.Zero();
        this.ikRotation = Quaternion.Identity();
    }
}

File: src/Runtime/Optimized/wasm_src/src/ik_chain_info.rs
use glam::{Quat, Vec3A};

pub(crate) struct IkChainInfo {
    local_rotation: Quat,
    local_position: Vec3A,
    ik_rotation: Quat,
}

impl IkChainInfo {
    pub(crate) fn new() -> Self {
        IkChainInfo {
            local_rotation: Quat::IDENTITY,
            local_position: Vec3A::ZERO,
            ik_rotation: Quat::IDENTITY,
        }
    }

    pub(crate) fn reset_state(&mut self) {
        self.local_rotation = Quat::IDENTITY;
        self.local_position = Vec3A::ZERO;
        self.ik_rotation = Quat::IDENTITY;
    }

    #[inline]
    pub(crate) fn local_rotation(&self) -> Quat {
        self.local_rotation
    }

    #[inline]
    pub(crate) fn local_rotation_mut(&mut self) -> &mut Quat {
        &mut self.local_rotation
    }

    #[inline]
    pub(crate) fn local_position(&self) -> Vec3A {
        self.local_position
    }

    #[inline]
    pub(crate) fn local_position_mut(&mut self) -> &mut Vec3A {
        &mut self.local_position
    }

    #[inline]
    pub(crate) fn ik_rotation(&self) -> Quat {
        self.ik_rotation
    }

    #[inline]
    pub(crate) fn ik_rotation_mut(&mut self) -> &mut Quat {
        &mut self.ik_rotation
    }
}
